### PR TITLE
Implement widget for the current speed

### PR DESCRIPTION
**Pull request by TobiPeterG**
Depends on https://github.com/organicmaps/organicmaps/pull/9355

This is a first draft to add a "widget" to show the current speed. It connects to the current speed limit and allows us to remove the speed from the bottom bar entirely. I'm completely happy with the current design, as it is quite distached from other UI elements at the moment. This is due to the next-turn-image that is sometimes visible and sometimes not. Maybe I should place it further down?

Also, the position in landscape mode is not optimal next to the lanes, it looks quite cluttered. It could be moved to the left side as well, but there it is blocked by the bottom menu (could be moved to the right) and two other buttons are also on that side interfering (could also be moved to the right side).

What are the thoughts about this?
The speed limit sign and connecting background appears/disappears with an animation.

### CODE REVIEW

**Comment by biodranik** [comment]
Does this implementation have any noticeable battery impact compared to the current master?

**Comment by TobiPeterG** [comment]
> Does this implementation have any noticeable battery impact compared to the current master?

Good question, didn't check. Will do :)

**Comment by TobiPeterG** [comment]
> Does this implementation have any noticeable battery impact compared to the current master?

So I did a quick test today, driving a route one way with this PR and one with the official playstore version
With my PR the battery dropped from 70 to 63 percent, on the playstore version it then dropped from 41 to 35 percent.
This is not very scientific, but it was about the same time, same distance and same phone. So a 1% difference at max, if this PR makes up the entire difference, but this isn't realistic . So I think the battery difference is neglectable, but more tests are of course appreciated :)

**Comment by TobiPeterG** [comment]
I fixed some issues, it works quite nicely :)
The changes I made:
I moved all map button to the right screen side, so they do not interfere with the widget. I moved the widget down on the right side, trying to have the same size as the map buttons.
The speed limit is now shown above the current speed, not below. This way the current speed is nicely alligned with the other elements. In horizontal mode, however, the speed limit is shown on the right side of the current speed due to space reasons.
Fixed some bugs:
the current speed widget disappeared when one had 0 km/h, not really expected, now it stays visible
the "alert" state when one drives faster than the speed limit used m/s instead of the local unit, meaning that it could happen that one goes slightly above the speed limit, but km/h still shows e.g. 50 in a 50 zone, but it turned red. I fixed it by using the localized strings, though I need to convert them back to int and this is really nastly, so if there's a better way, let me know :)



I found an unrelated issue though, the planning layout isn't used in the emulator when planning a route for the map buttons in horizontal layout. On my phone it works though (but the zoom buttons sometimes disappear?)

A related issue: Since I moved the widget to the bottom left, I now have to rely on the margin to the bottom to be correct. However, a "real" issue is that the bottom app bar moves up a bit to get out of the way of the system navigation bar. My speed widget doesn't follow though, the buttons do. I guess this code is responsible for this:
`MapButtonsController.java`
```
    @Override
    public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft,
                               int oldTop, int oldRight, int oldBottom)
    {
      mContentHeight = bottom - top;
      mContentWidth = right - left;
      mMapButtonsViewModel.setBottomButtonsHeight(getBottomButtonsHeight());
      mContentView.removeOnLayoutChangeListener(this);
    }

  private boolean isBehindPlacePage(View v)
  {
    if (mPlacePageViewModel == null)
      return false;
    final Integer placePageWidth = mPlacePageViewModel.getPlacePageWidth().getValue();
    if (placePageWidth != null)
      return !(mContentWidth / 2 > (placePageWidth.floatValue() / 2.0) + v.getWidth());
    return true;
  }

  private boolean isMoving(View v)
  {
    return v.getTranslationY() < 0;
  }

  public void move(float translationY)
  {
    if (mContentHeight == 0)
      return;

    // Move the buttons containers to follow the place page
    if (mInnerRightButtonsFrame != null &&
        (isBehindPlacePage(mInnerRightButtonsFrame) || isMoving(mInnerRightButtonsFrame)))
      applyMove(mInnerRightButtonsFrame, translationY);
    if (mInnerLeftButtonsFrame != null &&
        (isBehindPlacePage(mInnerLeftButtonsFrame) || isMoving(mInnerLeftButtonsFrame)))
      applyMove(mInnerLeftButtonsFrame, translationY);
  }

  private void applyMove(View frame, float translationY)
  {
    final float rightTranslation = translationY - frame.getBottom();
    final float appliedTranslation = rightTranslation <= 0 ? rightTranslation : 0;
    frame.setTranslationY(appliedTranslation);
    updateButtonsVisibility(appliedTranslation, frame);
  }

  public void updateButtonsVisibility()
  {
    if (mInnerLeftButtonsFrame != null)
      updateButtonsVisibility(mInnerLeftButtonsFrame.getTranslationY(), mInnerLeftButtonsFrame);
    if (mInnerRightButtonsFrame != null)
      updateButtonsVisibility(mInnerRightButtonsFrame.getTranslationY(), mInnerRightButtonsFrame);
  }
```

Is this correct? @vng could you have a look (btw thanks for your help :) )

**Comment by TobiPeterG** [comment]
If this is approved in its current form, I could remove the current speed from the bottom bar, so you'd have space for something new there :)

New video (oh brother, it doesn't like rotating the screen, but you get the idea ;) ):

https://github.com/user-attachments/assets/8d448a54-9d78-4ab6-bb10-403d4d4cae09

**Comment by TobiPeterG** [comment]
@biodranik I would like to add a speed limit sign for the German Autobahn when the speed limit is "unlimited", shown in OpenStreetMaps with maxspeed=none
Is this already mapped to the numeric speed limit value in Organic Maps somehow?

**Comment by vng** [comment]
How does the speed go to Android from the Core?
If it is FollowingInfo info, then check ```info.m_speedLimitMps == 0```
```
  if (speedLimit.IsNumeric())
    info.m_speedLimitMps = measurement_utils::KmphToMps(speedLimit.GetSpeedKmPH());
  else if (speedLimit.GetSpeed() == kNoneMaxSpeed)
    info.m_speedLimitMps = 0;
  else
    info.m_speedLimitMps = -1.0;
```

**Comment by TobiPeterG** [comment]
> How does the speed go to Android from the Core? If it is FollowingInfo, then check `info.m_speedLimitMps == 0`
> 
> ```
>   if (speedLimit.IsNumeric())
>     info.m_speedLimitMps = measurement_utils::KmphToMps(speedLimit.GetSpeedKmPH());
>   else if (speedLimit.GetSpeed() == kNoneMaxSpeed)
>     info.m_speedLimitMps = 0;
>   else
>     info.m_speedLimitMps = -1.0;
> ```

This worked, thank you :)

**Comment by TobiPeterG** [comment]
Alright, we now have a working sign for unlimited speed :)
It is the German sign (as it is the only industrial country where there is no speed limit i ncertain situations) and it works quite well :)
I also improved the appearance a bit, though I could only test it in the emulator yet. But so far, I'm very happy with the result :)
Next up is moving the widget up like the map buttons do and to remove the current speed from the bottom bar.
Any other wishes? :)

Video of the current version:

https://github.com/user-attachments/assets/ec28e761-98dd-471c-adc5-25b87839d68b

**Comment by TobiPeterG** [comment]
Oh btw. can someone give me a hint how I detect whether the system is in dark mode or light mode?

**Comment by foss-** [comment]
Had not followed this closely, so just a question: in the last video the current speed changes *a lot*, frequently - sometimes multiple times per second. If the video is in real-time, I find this very hectic, hard to follow and distracting.

Maybe this is a feature request for a follow-up PR, but would it make sense to use averages of current speed of e.g. the last few seconds, rather than just showing any current speed? While that is a bit less accurate it would be a lot less distracting and hectic.

**Comment by TobiPeterG** [comment]
> Had not followed this closely, so just a question: in the last video the current speed changes _a lot_ - sometimes multiple times per second. If the video is in real-time, I find this very hectic, hard to follow and distracting.
> 
> Maybe this is a feature request for a follow-up PR, but would it make sense to use averages of current speed of e.g. the last few seconds, rather than just showing any current speed? While that is a bit less accurate it would be a lot less distracting and hectic.

Thanks for the comment :)
The hectic changes happen because the emulator is not very good in emulating consistent speed, instead, it just moves the current position in small steps - leading to jumps in the current speed, this doesn't happen in a real-world scenario - I just can't drive on an Autobahn with my car every time I want to test a small change ;)

I think the speed is updated every second. Feel free to test this PR and check if it disturbes you in real-life and then we can explore alternatives :)

**Comment by gpesquero** [comment]
Hi @TobiPeterG. First of all, great work with this is PR!!

I've been testing it and the main issue that I've detected is regarding the widget position in landscape mode. 

In the Android emulator, this widget is partially hidden by the navigation bar:

imagen_000

And in my real device, the widget is too close (touching) the navigation bar:

imagen_001

And if I flip the device 180 degrees so the Android system bar is in the left side, the widget is way on the left, under the system bar:

imagen_002

I'm no expert with Android layout issues. If it's any help, I could "advertise" this PR in the Contributors' UX Telegram channel so maybe someone can give us a hint how to solve this issue.

**Comment by TobiPeterG** [comment]
> Hi @TobiPeterG. First of all, great work with this is PR!!
> 
> I've been testing it and the main issue that I've detected is regarding the widget position in landscape mode. 
> 
> In the Android emulator, this widget is partially hidden by the navigation bar:
> 
> imagen_000
> 
> And in my real device, the widget is too close (touching) the navigation bar:
> 
> imagen_001
> 
> And if I flip the device 180 degrees so the Android system bar is in the left side, the widget is way on the left, under the system bar:
> 
> imagen_002
> 
> I'm no expert with Android layout issues. If it's any help, I could "advertise" this PR in the Contributors' UX Telegram channel so maybe someone can give us a hint how to solve this issue.


Hey, thanks for testing :)
Yes, I'm aware of the issue, I mentioned it above already.
The other UI elements have an extra function to move them away from the system bars. I'll "just" have to adapt that as well. :)

**Comment by gpesquero** [comment]
> Oh btw. can someone give me a hint how I detect whether the system is in dark mode or light mode?

You can use the static call `Framework.nativeGetMapStyle()` to get the current map light/dark style as defined in `map_style.hpp` file:
```
  MapStyleDefaultLight = 0,
  MapStyleDefaultDark = 1,
  MapStyleMerged = 2,
  MapStyleVehicleLight = 3,
  MapStyleVehicleDark = 4,
  MapStyleOutdoorsLight = 5,
  MapStyleOutdoorsDark = 6,
```

**Comment by gpesquero** [comment]
@TobiPeterG: IMO the size of the texts for the speed value and units in the current implementation are a little bit small... I'd set the text size for the value at least the same size as the speed text size of the navigation bar:

imagen_003

**Comment by TobiPeterG** [comment]
> @TobiPeterG: IMO the size of the texts for the speed value and units in the current implementation are a little bit small... I'd set the text size for the value at least the same size as the speed text size of the navigation bar:
> 
> imagen_003

Thank you for the feedback :)

I accept this concern, however, I couldn't find a better solution (yet)
I wanted the speed widget to have the same size as the map buttons on the right side for consistency reasons. Secondly, I didn't want the text to touch the border of the widget or be too close to it.
That's why it's quite small, I think this as well. Do you have an idea how it can be improved without breaking one of the above concerns?

**Comment by gpesquero** [comment]
> I wanted the speed widget to have the same size as the map buttons on the right side for consistency reasons. Secondly, I didn't want the text to touch the border of the widget or be too close to it.
> That's why it's quite small, I think this as well. Do you have an idea how it can be improved without breaking one of the above concerns?

No, I don't think that we can keep the same size as the map buttons and have bigger texts at the same time.

Here's a screenshot of how it looks on the emulator making the widget bigger so the text is similar to the one in the navigation bar. At a first glance, it looks too big to me, but I will test it in the next days on a real device to check if it's really that big.

Captura desde 2024-12-08 23-24-57

**Comment by Misalf-git** [comment]
In the screenshots, the digits and the "unit of speed" (km/h) _in combination_ are centered vertically inside the circle, resulting in the digits to seem shifted upwards.

If only the digits are centered vertically while ignoring "unit of speed" from any centering (i.e. display units below digits like maybe by downwards offset) then there will be more space to the circle's inner border (left/right), making the digits look less cramped.

**Comment by TobiPeterG** [comment]
> > I wanted the speed widget to have the same size as the map buttons on the right side for consistency reasons. Secondly, I didn't want the text to touch the border of the widget or be too close to it.
> > That's why it's quite small, I think this as well. Do you have an idea how it can be improved without breaking one of the above concerns?
> 
> No, I don't think that we can keep the same size as the map buttons and have bigger texts at the same time.
> 
> Here's a screenshot of how it looks on the emulator making the widget bigger so the text is similar to the one in the navigation bar. At a first glance, it looks too big to me, but I will test it in the next days on a real device to check if it's really that big.
> 
> Captura desde 2024-12-08 23-24-57

I adjusted the code to center the current speed and have the units below uncentered. This way I could make the text a bit bigger:
Screenshot_20241221_145412

In my opinion, this is perfectly fine. The new text has 83% of the size of the text in the bottom bar (50 vs 60 pixels on my device). Is this alright?

**Comment by TobiPeterG** [comment]
I added "dark mode support", for now only the background of the two widgets. IMO already looking good, what do you guys think? Should the current speed widget also get a dark background?
Screenshot_20241221_152857

**Comment by TobiPeterG** [comment]
The widget now isn't overlapping anymore :partying_face: 
Please test. I noticed that it is still not alligned with the buttons on the right, I don't know what magic is happening on their side.

image

**Comment by TobiPeterG** [comment]
Removed the current speed from the bottom bar
image

I noticed that this also removes the "warning" when near a speed camera, how bad is this? I would like to add a proper warning badge later on anyway.

**Comment by TobiPeterG** [comment]
Now also working with a navigation bar on the left side ;)
image

**Comment by TobiPeterG** [comment]
Are there any other wishes? :)
Otherwise, I think this is ready :)

**Comment by Jean-BaptisteC** [comment]
Search and bookmarks buttons are hide in landscape

File: android/app/src/main/res/values/attrs.xml [hunk 1]
```diff
@@ -2,7 +2,9 @@
 <resources>
   <declare-styleable name="SpeedLimitView">
     <attr name="BackgroundColor" format="color" />
+    <attr name="WidgetBackgroundColor" format="color" />
```
Comment (AndrewShkrob) [inline android/app/src/main/res/values/attrs.xml:5]: What's the difference between `BackgroundColor` and `WidgetBackgroundColor`?
It should be possible to understand it from the name.
Btw, why do these values have names starting from upper case and values below have names starting from lower case?
They all should start from lower case letters like any other android's XML value.

File: android/app/.attach_pid35553
Comment (AndrewShkrob) [inline android/app/.attach_pid35553:1]: Remove the file

File: android/app/src/main/java/app/organicmaps/widget/SpeedLimitView.java [hunk 2]
```diff
@@ -94,110 +147,349 @@ public SpeedLimitView(Context context, @Nullable AttributeSet attrs)
...
+    if (mCurrentSpeedMps < 0)
+      mCurrentSpeedStr = "0";
+    else
+      mCurrentSpeedStr = speedAndUnits.first;
+    mCurrentSpeedUnitsStr = speedAndUnits.second;
+    configureTextSize();
     invalidate();
   }
 
+  private void draw_widgets(@NonNull Canvas canvas, float cx, float cx_or_cy, float cy, boolean isLandscape) {
```
Comment (AndrewShkrob) [inline android/app/src/main/java/app/organicmaps/widget/SpeedLimitView.java:249]:
```suggestion
  private void drawWidgets(@NonNull Canvas canvas, float cx, float cx_or_cy, float cy, boolean isLandscape) {
```

File: android/app/src/main/java/app/organicmaps/widget/SpeedLimitView.java [hunk 3]
```diff
@@ -207,15 +499,15 @@ private void configureTextSize()
     while (lowerBound <= upperBound)
     {
       textSize = (lowerBound + upperBound) / 2;
-      mTextPaint.setTextSize(textSize);
-      mTextPaint.getTextBounds(text, 0, text.length(), textBounds);
+      textPaint.setTextSize(textSize);
+      textPaint.getTextBounds("180", 0, 3, textBounds);
 
-      if (Math.pow(textBounds.width(), 2) + Math.pow(textBounds.height(), 2) <= textMaxSizeSquared)
+      if (textBounds.width() * textBounds.width() + textBounds.height() * textBounds.height() <= textMaxSizeSquared)
```
Comment (AndrewShkrob) [inline android/app/src/main/java/app/organicmaps/widget/SpeedLimitView.java:505]: What's wrong with `Math.pow()`? Does your solution work faster?

File: android/app/src/main/java/app/organicmaps/widget/menu/NavMenu.java [hunk 4]
```diff
@@ -213,20 +206,6 @@ private void updateSpeedView(@NonNull RoutingInfo info)
...
-      if (info.isSpeedCamLimitExceeded())
-        mSpeedValue.setTextColor(ContextCompat.getColor(mActivity, R.color.white_primary));
-      else
-        mSpeedValue.setTextColor(ContextCompat.getColor(mActivity, R.color.base_red));
-    }
-    else
-      mSpeedValue.setTextColor(ThemeUtils.getColor(mActivity, android.R.attr.textColorPrimary));
-
-    mSpeedUnits.setText(speedAndUnits.second);
-    mSpeedViewContainer.setActivated(info.isSpeedCamLimitExceeded());
```
Comment (AndrewShkrob) [inline android/app/src/main/java/app/organicmaps/widget/menu/NavMenu.java:229]: Nobody approved this removal yet :)
I think it's better to make it in separate PR

**Review by AndrewShkrob — CHANGES_REQUESTED** [review]
Unfortunately, there are a lot of things I don't like in this PR :(
1. As I already told you before, please format the code with our formatter (and don't forget about empty lines between methods): https://github.com/organicmaps/organicmaps/pull/9355#issuecomment-2370809004
2. I don't see the point in performing a code review until the main architectural issue is resolved.

Let's discuss the 2nd point.
Imagine you're a developer that sees this code for the first time. We have a class `SpeedLimitView`. What do you think this class does?
You would probably say something like:
- Hey, the class name has a `SpeedLimit` and a `View`. I would assume it draws a speed limit sign on the screen.
Well, you would be surprised when you discover that instead of a simple speed limit sign it also draws a current speed, unlimited sign, a background to combine all of these signs together and it does much more stuff under the hood.

Let's try to split the code into separate classes to make it maintainable and easily extendable in the future.

How to do this?
The reason why you decided to write all the code inside the one class is because the code is pretty similar however it is a different functionality.
So, let's move all the similar code into a base class (f.e. `SignView`).
Then for each specific task (in our case for each specific sign) we will have a separate class - `SpeedLimitView`, `CurrentSpeedView`, `UnlimitedSignView`. 
This way it would be easily maintainable because each class will have a lot less lines of code and it would be easily extendable f.e. when we need to draw another sign we'll just create another class instead of diving inside a class with over 1k lines of code trying to find a place where to add our new code and later spending week in testing that we didn't break anything :)

I don't like talking about OOP and patterns and all this stuff but looks like it's a good time :)
Please, check SOLID principles. SRP and OCP are exactly what your code needs right now.

That was the first main issue. The second one is how to combine all these signs.
We simply need a container class here.
I have some thoughts:
1. Have you considered using LinearLayout for this? You can configure background and animations for it in XML. No need to write the code especially for such simple things.
2. If this doesn't work then you can consider extending LinearLayout class with your own functionality.
3. If this also doesn't work then (and only then) we can think about writing our custom container class that will hold and place our sign views in the correct places.

Very important note.
We're discussing here UI classes. They MUST behave like other Android's UI elements. All these classes (SpeedLimitView, CurrentSpeedView, UnlimitedSignView, SignContainer) must work like default android views. You should be able to configure them in layout inspector, to see their dimensions/sizes, where they are placed, how they behave with different parameters and so on. And, of course, 
you must be able to see them rendered in the layout inspector.

After these issues are resolved, we can proceed with the code review. Until then, sorry, I can't approve it.

**Comment by TobiPeterG** [comment]
@AndrewShkrob 

Thank you for your feedback :)
I guess I should have been clearer - with "ready" I didn't mean "ready to be merged", but ready that someone has a look at it that
A) the required features are implemented 
And 
B) a superficial look at the architecture

Sorry if my message was misleading.

I don't have much experience in Android development _yet_, so I didn't even attempt to make it right code wise. I wanted to first create a prototype to show that what I envision can work and to find out if users/devs are even interested in this. And if yes (which seems to be the case), think about how to do it right (and you have provided a lot of ideas on how to do it right, so thanks :) )
However, there are still a few issues left, for example, the widget is not aligned on the y axis with the elements on the right (mentioned above). Do you know why this happens?
I wanted to prevent implementing something "the right way" just to find out that I missed something that will require a major overhaul.
I'm well aware of the shortcomings and inconsistencies of my current code. :)

File: android/app/src/main/java/app/organicmaps/widget/menu/NavMenu.java (hunk 4 above)
Comment (gpesquero) [reply android/app/src/main/java/app/organicmaps/widget/menu/NavMenu.java:229]: > Nobody approved this removal yet :)
> I think it's better to make it in separate PR

@AndrewShkrob, @TobiPeterG: I've launched [PR #9955](https://github.com/organicmaps/organicmaps/pull/9955) for the removal of the speed display in the navigation panel, so we can discuss there the rearrange of the items in the navigation panel while the new Speed widget is implemented.

**Comment by gpesquero** [comment]
> Removed the current speed from the bottom bar [ image ]
> I noticed that this also removes the "warning" when near a speed camera (aka. the background changes to red, without telling you why), how bad is this? I would like to add a proper warning badge later on anyway.

@TobiPeterG: Now that the speed cam warning will disappear from the bottom navigation panel, my proposal would be to display a "blinking" speed cam icon on top of the speed limit display. Something similar like this:

speed_cam

Here are the base icons:
imagen_004a   imagen_004b   imagen_004d

**Comment by AndrewShkrob** [comment]
PTAL at CarPlay implementation: https://github.com/organicmaps/organicmaps/issues/2968

**Comment by soshial** [comment]
Thank you so much for implementing this, @TobiPeterG!

May I ask to implement this in Android Auto as well? Currently, now speed limit is shown there.
